nil/internal/collate/proposer.go

```diff
@@ -150,6 +161,80 @@ func (p *proposer) fetchLastBlockHashes(tx db.RoTx) error {
 	return nil
 }
 
+func (p *proposer) handleL1Attributes(tx db.RoTx) error {
+	if !p.params.ShardId.IsMainShard() {
+		return nil
+	}
+	if p.l1BlockFetcher == nil {
+		return errors.New("L1 block fetcher is not initialized")
+	}
+
+	block, err := p.l1BlockFetcher.GetLastBlockInfo(p.ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get latest L1 block: %w", err)
+	}
+	if block == nil {
+		// No block yet
+		return nil
+	}
+
+	// Check if this L1 block was already processed
+	if cfgAccessor, err := config.NewConfigReader(tx, nil); err == nil {
+		if prevL1Block, err := config.GetParamL1Block(cfgAccessor); err == nil {
+			if prevL1Block != nil && prevL1Block.Number >= block.Number.Uint64() {
+				return nil
+			}
+		}
+	}
+
+	txn, err := CreateL1BlockUpdateTransaction(block)
+	if err != nil {
+		return fmt.Errorf("failed to create L1 block update transaction: %w", err)
+	}
+
+	p.logger.Trace().Msgf("Add L1 block update transaction")
```

Comment:
Maybe let's add message hash to log entry?
```suggestion
	p.logger.Trace().Msg("Add L1 block update transaction")
```